File: reisepilot-build/project/app/src/main/java/de/balabucha/reisepilot/MainActivity.java
package de.balabucha.reisepilot;

import android.Manifest;
import android.app.Activity;
import android.content.*;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.*;

public class MainActivity extends Activity {
    private static final int REQ_LOCATION = 100;
    private String pendingTrip = null;
    private TextView status;
    private TextView metrics;

    private final BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override public void onReceive(Context context, Intent intent) {
            String text = intent.getStringExtra("text");
            if (text != null) metrics.setText(text);
        }
    };

    @Override protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(buildUi());
        IntentFilter f = new IntentFilter(TripTrackingService.ACTION_UPDATE);
        if (Build.VERSION.SDK_INT >= 33) registerReceiver(receiver, f, RECEIVER_NOT_EXPORTED);
        else registerReceiver(receiver, f);
        if (Build.VERSION.SDK_INT >= 33 && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.POST_NOTIFICATIONS}, 101);
        }
    }

    private View buildUi() {
        ScrollView scroll = new ScrollView(this);
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(30, 30, 30, 50);
        root.setBackgroundColor(Color.rgb(245,247,250));
        scroll.addView(root);

        TextView title = text("ReisePilot", 30, true);
        root.addView(title);
        root.addView(text("Schwerin → Montbéliard → Canet", 16, false));
        root.addView(space(18));

        status = text("Noch nicht gestartet.", 18, true);
        status.setPadding(20,20,20,20);
        status.setBackgroundColor(Color.rgb(255,244,219));
        root.addView(status);

        metrics = text("Abfahrt Samstag: 07:30–08:00 Uhr\nStandard: 07:45 Uhr", 18, false);
        metrics.setPadding(0,20,0,20);
        root.addView(metrics);

        root.addView(button("Samstag starten", v -> startTrip("sat")));
        root.addView(button("Sonntag starten", v -> startTrip("sun")));
        root.addView(button("Pause / Weiter", v -> sendService(TripTrackingService.PAUSE)));
        root.addView(button("Tracking stoppen", v -> sendService(TripTrackingService.STOP)));
        root.addView(space(18));

        root.addView(text("Navigation", 22, true));
        root.addView(button("Google Maps: Schwerin → Hotel", v -> openMaps("sat")));
        root.addView(button("Google Maps: Hotel → Canet", v -> openMaps("sun")));
        root.addView(button("Coyote öffnen / installieren", v -> launchPackage("com.coyotesystems.android")));
        root.addView(space(18));

        root.addView(text("Pausen und Tanken", 22, true));
        root.addView(text("Samstag 10:15 kurze Pause\n13:00 Pause + günstiger Tankstopp direkt an der Route\n16:30 zweite Pause\n20:00–22:00 greet Hôtel\n\nSonntag 07:00 Frühstück\n07:35–07:45 Abfahrt\n12:45 Pause + Tanken Intermarché Orange\n17:00–19:00 Malibu Village", 17, false));
        root.addView(space(18));

        root.addView(text("Mautpunkte", 22, true));
        root.addView(text("A36 Fontaine-Larivière\nA36 Saint-Maurice / Écot\nA7 Vienne\nA9 Perpignan Nord\n\nDie App warnt ab ungefähr 20 km Entfernung.", 17, false));
        root.addView(space(18));

        root.addView(text("Kontakte", 22, true));
        root.addView(button("greet Hôtel anrufen", v -> dial("+33381901069")));
        root.addView(button("Malibu Village anrufen", v -> dial("+33468732779")));
        root.addView(text("Malibu Check-in 16:00–19:00 Uhr; nach vorherigem Anruf bis 23:00 Uhr.", 16, false));
        return scroll;
    }

    private TextView text(String value, int sp, boolean bold) {
        TextView v = new TextView(this);
        v.setText(value);
        v.setTextSize(sp);
        v.setTextColor(Color.rgb(23,32,51));
        if (bold) v.setTypeface(null, android.graphics.Typeface.BOLD);
        return v;
    }

    private View space(int dp) {
        Space s = new Space(this);
        s.setLayoutParams(new LinearLayout.LayoutParams(1, dp));
        return s;
    }

    private Button button(String label, View.OnClickListener listener) {
        Button b = new Button(this);
        b.setText(label);
        b.setAllCaps(false);
        b.setOnClickListener(listener);
        LinearLayout.LayoutParams p = new LinearLayout.LayoutParams(-1, -2);
        p.setMargins(0, 8, 0, 8);
        b.setLayoutParams(p);
        return b;
    }

    private boolean hasLocation() {
        return checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
               checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;
    }

    private void startTrip(String trip) {
        if (!hasLocation()) {
            pendingTrip = trip;
            requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION}, REQ_LOCATION);
            return;
        }
        status.setText("Tracking wird gestartet …");
        Intent i = new Intent(this, TripTrackingService.class);
        i.setAction("sun".equals(trip) ? TripTrackingService.START_SUN : TripTrackingService.START_SAT);
        startForegroundService(i);
    }

    private void sendService(String action) {
        Intent i = new Intent(this, TripTrackingService.class);
        i.setAction(action);
        startService(i);
    }

    @Override public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQ_LOCATION && hasLocation() && pendingTrip != null) {
            String trip = pendingTrip;
            pendingTrip = null;
            startTrip(trip);
        }
    }

    private void openMaps(String trip) {
        String url = "sun".equals(trip)
            ? "https://www.google.com/maps/dir/?api=1&origin=greet+H%C3%B4tel+Montb%C3%A9liard&destination=Malibu+Village+Canet-en-Roussillon&travelmode=driving"
            : "https://www.google.com/maps/dir/?api=1&origin=19057+Schwerin&destination=greet+H%C3%B4tel+Montb%C3%A9liard&travelmode=driving";
        startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(url)));
    }

    private void launchPackage(String pkg) {
        Intent launch = getPackageManager().getLaunchIntentForPackage(pkg);
        if (launch != null) startActivity(launch);
        else startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + pkg)));
    }

    private void dial(String number) {
        startActivity(new Intent(Intent.ACTION_DIAL, Uri.parse("tel:" + number)));
    }

    @Override protected void onDestroy() {
        unregisterReceiver(receiver);
        super.onDestroy();
    }
}

File: reisepilot/app/src/main/java/de/balabucha/reisepilot/MainActivity.kt
package de.balabucha.reisepilot

import android.Manifest
import android.content.*
import android.content.pm.PackageManager
import android.net.Uri
import android.os.*
import android.view.WindowManager
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.runtime.*
import androidx.core.content.ContextCompat

class MainActivity : ComponentActivity() {
    companion object {
        private const val STARTUP_LOCATION_REQUEST = 5601
    }

    private var snapshot by mutableStateOf(TripSnapshot())
    private lateinit var realtime54: RealtimeDriveController54
    private lateinit var nearby56: NearbyLiveController56

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            intent?.getStringExtra("snapshot")?.let {
                snapshot = TripSnapshot.fromJson(it)
                if (::realtime54.isInitialized) realtime54.onSnapshot(snapshot)
                if (::nearby56.isInitialized) nearby56.onSnapshot(snapshot)
            }
            if (intent?.getBooleanExtra("appsChanged", false) == true) {
                snapshot = snapshot.copy(lastUpdatedEpochMs = System.currentTimeMillis())
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        loadSnapshot()
        realtime54 = RealtimeDriveController54(this)
        nearby56 = NearbyLiveController56(this)
        realtime54.onSnapshot(snapshot)
        nearby56.onSnapshot(snapshot)
        ContextCompat.registerReceiver(
            this,
            receiver,
            IntentFilter(TripTrackingService.ACTION_UPDATE),
            ContextCompat.RECEIVER_NOT_EXPORTED
        )
        setContent { ReiseTheme { ReisePilotApp(this, snapshot) } }
        maybeRequestStartupLocation()
    }

    override fun onResume() {
        super.onResume()
        loadSnapshot()
        if (::realtime54.isInitialized) {
            realtime54.onSnapshot(snapshot)
            realtime54.onResume()
        }
        if (::nearby56.isInitialized) {
            nearby56.onSnapshot(snapshot)
            nearby56.onResume()
        }
    }

    override fun onPause() {
        if (::realtime54.isInitialized) realtime54.onPause()
        if (::nearby56.isInitialized) nearby56.onPause()
        super.onPause()
    }

    override fun onDestroy() {
        unregisterReceiver(receiver)
        if (::realtime54.isInitialized) realtime54.destroy()
        if (::nearby56.isInitialized) nearby56.destroy()
        super.onDestroy()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == STARTUP_LOCATION_REQUEST && ::nearby56.isInitialized) {
            nearby56.onPermissionChanged()
        }
    }

    private fun maybeRequestStartupLocation() {
        val granted = checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
            checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED
        if (granted) return
        val prefs = getSharedPreferences("settings", MODE_PRIVATE)
        if (prefs.getBoolean("startup_live_location_prompted_v56", false)) return
        prefs.edit().putBoolean("startup_live_location_prompted_v56", true).apply()
        requestPermissions(
            arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION),
            STARTUP_LOCATION_REQUEST
        )
    }

    private fun loadSnapshot() {
        val loaded = TripSnapshot.fromJson(
            getSharedPreferences("trip_state", MODE_PRIVATE).getString("snapshot", null)
        )
        snapshot = if (
            !loaded.active && loaded.tripMode == TripMode.TEST &&
            tripModeAt(java.time.Instant.now()) == TripMode.REAL
        ) {
            TripSnapshot(
                stage = loaded.stage,
                tripMode = TripMode.REAL,
                nextTitle = "Reise bereit",
                nextDetail = "Reise starten"
            ).also {
                getSharedPreferences("trip_state", MODE_PRIVATE).edit()
                    .putString("snapshot", it.json().toString())
                    .apply()
            }
        } else {
            loaded
        }
    }

    fun startTrip(stage: Stage) {
        val action = if (stage == Stage.SATURDAY) {
            TripTrackingService.ACTION_START_SAT
        } else {
            TripTrackingService.ACTION_START_SUN
        }
        ContextCompat.startForegroundService(
            this,
            Intent(this, TripTrackingService::class.java).setAction(action)
        )
    }

    fun serviceAction(action: String) {
        startService(Intent(this, TripTrackingService::class.java).setAction(action))
    }

    fun reloadTripConfig() {
        serviceAction(TripTrackingService.ACTION_RELOAD_CONFIG)
    }

    fun refreshRoadAhead54() {
        if (snapshot.active) {
            if (::realtime54.isInitialized) realtime54.forceRefresh()
        } else {
            if (::nearby56.isInitialized) nearby56.forceRefresh()
        }
    }

    fun updateKeepScreenOn(tripActive: Boolean) {
        val enabled = getSharedPreferences("settings", MODE_PRIVATE)
            .getBoolean("keep_screen_on", true) && tripActive
        if (enabled) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        } else {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    fun updateApiStatus(ok: Boolean, message: String) {
        snapshot = snapshot.copy(apiOk = ok, apiMessage = message)
        getSharedPreferences("trip_state", MODE_PRIVATE).edit()
            .putString("snapshot", snapshot.json().toString())
            .apply()
    }

    fun openMaps(stage: Stage) {
        val uri = if (stage == Stage.SATURDAY) {
            "https://www.google.com/maps/dir/?api=1&origin=19057+Schwerin&destination=greet+H%C3%B4tel+Montb%C3%A9liard&travelmode=driving"
        } else {
            "https://www.google.com/maps/dir/?api=1&origin=greet+H%C3%B4tel+Montb%C3%A9liard&destination=Malibu+Village+Canet-en-Roussillon&travelmode=driving"
        }
        startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(uri)))
    }

    fun openPointRoute(point: GeoPoint, label: String) {
        val destination = "${point.lat},${point.lon}"
        val uri = Uri.parse(
            "https://www.google.com/maps/dir/?api=1&destination=${Uri.encode(destination)}" +
                "&destination_place_id=&travelmode=driving"
        )
        startActivity(Intent(Intent.ACTION_VIEW, uri).apply {
            putExtra("query", label)
        })
    }

    fun openMapSearch(query: String) {
        startActivity(
            Intent(
                Intent.ACTION_VIEW,
                Uri.parse("https://www.google.com/maps/search/?api=1&query=${Uri.encode(query)}")
            )
        )
    }

    fun openWeb(url: String) {
        if (url.startsWith("https://")) {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
        }
    }

    fun isInstalled(pkg: String): Boolean = try {
        packageManager.getPackageInfo(pkg, 0)
        true
    } catch (_: Exception) {
        false
    }

    fun openPackage(pkg: String) {
        val launch = packageManager.getLaunchIntentForPackage(pkg)
        if (launch != null) {
            startActivity(launch)
        } else {
            startActivity(
                Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("https://play.google.com/store/apps/details?id=$pkg")
                )
            )
        }
    }

    fun openBisonFute() {
        val pkg = "fr.gouv.bisonfute"
        val launch = packageManager.getLaunchIntentForPackage(pkg)
        if (launch != null) {
            startActivity(launch)
        } else {
            startActivity(
                Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("https://www.bison-fute.gouv.fr/information-trafic.html")
                )
            )
        }
    }

    fun dial(number: String) {
        startActivity(Intent(Intent.ACTION_DIAL, Uri.parse("tel:$number")))
    }
}
